[MXNET-304][RFC] Jenkins docs build

aaronmarkham opened the pull request:
## Description ##

Adapted updated build scripts to the new Jenkins CI setup. This enables nightly docs builds.
**Preview:** http://107.21.90.138/

Fixes #10238
Fixes #10226
Fixes #10273

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Trigger via Jenkins CI (currently pointing to my forks)
    1. Use [Pipeline website build - test pipeline](http://jenkins.mxnet-ci.amazon-ml.com/job/website%20build%20-%20test%20pipeline/): this uses parameters from the UI so you can select which tag(s) to build, which tags should appear in the versions dropdown, which tag to be default (now master), and the domain (to facilitate dev server testing).
    2. Then run [Project website build - test publish](http://jenkins.mxnet-ci.amazon-ml.com/job/website%20build%20-%20test%20publish/): this is mostly a clone from the previous site deploy Jenkins job; it deletes everything and grabs the artifacts from the build job and pushes those new files as a commit to the asf-site branch of mxnet-incubator-site.

- [x] Updated [dependencies](https://github.com/aaronmarkham/incubator-mxnet/blob/1c08ec9b3b93a0a108e62ac4c21dfec9c7002274/ci/docker/install/ubuntu_docs.sh) for the docs Docker image
- [x] Uses a [Jenkinsfile](https://github.com/aaronmarkham/incubator-mxnet/blob/1c08ec9b3b93a0a108e62ac4c21dfec9c7002274/docs/Jenkinsfile) and [runtime_functions.sh](https://github.com/aaronmarkham/incubator-mxnet/blob/1c08ec9b3b93a0a108e62ac4c21dfec9c7002274/ci/docker/runtime_functions.sh) to call the [build_version_doc/build_all_version.sh](https://github.com/aaronmarkham/incubator-mxnet/blob/1c08ec9b3b93a0a108e62ac4c21dfec9c7002274/docs/build_version_doc/build_all_version.sh) to build the desired tags, then calls [update_all_version.sh](https://github.com/aaronmarkham/incubator-mxnet/blob/1c08ec9b3b93a0a108e62ac4c21dfec9c7002274/docs/build_version_doc/update_all_version.sh) to inject a variety of changes to the site versions. These are then saved as [artifacts.tgz](http://jenkins.mxnet-ci.amazon-ml.com/job/website%20build%20-%20test%20pipeline/lastSuccessfulBuild/artifact/).
- [x] `update_all_version.sh` and `build_all_version.sh` now use IFS to delimit the tags by `;` since Jenkins has a hard time passing quotes downstream to the shell scripts.
    * This means usage for dev builds will change like this to use semicolons: `./update_all_version.sh master;1.1.0;1.0.0 master http://localhost`

## Usage ##
**How do you test this stuff?**
1. Login to [Jenkins](http://jenkins.mxnet-ci.amazon-ml.com).
2. View the pipeline currently called [website build - test pipeline](http://jenkins.mxnet-ci.amazon-ml.com/job/website%20build%20-%20test%20pipeline/)
3. Click **Build with Parameters**.
4. Use the defaults, or change the domain to be your staging server's IP/DNS web address.
5. Wait about 20-30 minutes while it builds the full site.
6. On your staging server, clone my [mxnet site repo](https://github.com/aaronmarkham/incubator-mxnet-site/).
    - When you ran [website build - test pipeline](http://jenkins.mxnet-ci.amazon-ml.com/job/website%20build%20-%20test%20pipeline/) it followed up with [website build -test publish](http://jenkins.mxnet-ci.amazon-ml.com/job/website%20build%20-%20test%20publish/) which pushed the changes to my fork of [incubator-mxnet-site](https://github.com/apache/incubator-mxnet-site).
    - Make sure you `git pull` if you already had a clone of my fork.
7. Copy the files to your webroot. For more info on this see the [developer instructions for docs build](https://github.com/apache/incubator-mxnet/tree/master/docs/build_version_doc#serving-your-development-version).
8. Preview the site on your staging server. Note, use the domain default before you try to use this for production, but using your own is helpful for QA'ing the site.

## Production Steps ##

After this PR is merged, both of the Jenkins jobs need to be updated to use production instead of my forks. Credentials for publishing to the production site repo will need to provided in the job's configuration. Note, don't change the names because the Jenkinsfile code for calling publish calls it by name.

## Comments ##
- The site deploy CI job should be converted to use Jenkinsfile instead of having script logic in the job itself. So that's a future TODO.
- The publish job shouldn't delete the whole site repo. You should be able to just run master for a docs build and replace just master on the site. Another future TODO.
- You should be able to run these scripts in the context of your current fork. At the moment they're hardcoded to checkout the project's main fork. This means that you can build the docs locally still with `make docs`, but you won't get to see all of the customizations that happen with the `update_all_version.sh` script. Another future TODO.
- I'm seeing hardcoded problems: for current tag as default and the domain. These should be ferreted out as we test this new build structure. The site should work on a dev server perfectly, without directing to a hardcoded url. (Too often during testing you realize you're not on your dev server anymore, but on production.)
- Artifacts could be handled better. We should investigate different plugins and artifacts strategies, so that we build smarter.
- There's some waste on the Docker deps - these should probably be baked into an image and loaded on docker hub so that the jobs aren't reinstalling this stuff every time.
- Jumping from the CI groovy script to shell is not the right pattern from what I can tell. Seems like these things should be refactored to be all in groovy and using built-in Jenkins functions. Another TODO.

## Notes ##
* Make use of the [Pipeline Syntax link](http://jenkins.mxnet-ci.amazon-ml.com/job/website%20build%20-%20test%20pipeline/pipeline-syntax/) at the bottom of the Jenkins configure screen for the jobs. Many buried features in there to help build your job.

marcoabreu commented on ci/docker/install/ubuntu_docs.sh:
```diff
@@ -21,8 +21,44 @@
+    curl \
+    doxygen \
+    git \
+    libatlas-base-dev \
```
Are all these dependencies really required? Installing them here might override other version restrictions. It would be better to run this script after the other install scripts and then see what's actually required for the document generation.

aaronmarkham replied:
I'd like to just bake this stuff into a docker image. But to answer the question: these are all of the requirements needed to build the docs. Except R. That takes more deps and isn't included in the required scope for the nightlies. It is possible that you can get *a result* without some of these, but I resolved warning after warning and error after error until I had a *viable* result.

Refer to: https://github.com/apache/incubator-mxnet/issues/10273
See render in staging:
http://107.21.90.138/api/python/ndarray/ndarray.html#mxnet.ndarray.Correlation

and
https://github.com/apache/incubator-mxnet/issues/10226
Fixed in staging:
http://107.21.90.138/versions/master/api/python/io/io.html#mxnet.io.ImageRecordIter

Also, this is the script for the ubuntu_docs, so why does overriding anything matter? It's for docs. It's what the docs need to be happy.

marcoabreu replied:
Overriding matters because this script is part of the ubuntu_cpu dockerfile pipeline (https://github.com/apache/incubator-mxnet/blob/master/ci/docker/Dockerfile.build.ubuntu_cpu#L45). 

I don't think we need to install libjemalloc-dev, liblapack-dev, libopenblas-dev, libopencv-dev, numpy, git and all the other libraries at this point since they should already be present in the underlying layers (see https://github.com/apache/incubator-mxnet/blob/master/ci/docker/install/ubuntu_core.sh#L27).

aaronmarkham replied:
None of these are versioned at any point. I'd argue that while layering makes sense, it opens things up to breakage when we don't know what the full stack is for docs. For example, one of my deps, `libatlas-base-dev ` is only appearing in the Caffe script. If scripts gets pulled out for some reason, then docs may break. 

`libjemalloc-dev` doesn't appear anywhere in the other scripts despite every build run telling you that you should have it installed (to massive improve speed, or so it contends).

These seem like they should move to the base script:
```
    apt-transport-https \
    ca-certificates \
    software-properties-common \
```

These would seem to not be relevant to CI, but I imagine if tutorials are being automatically executed then they're required, but at what layer? Just for docs?
```
    curl \
    unzip \
    wget
```

marcoabreu replied:
Right, at the moment they're not versioned. The idea here is to make sure every single dependency is only defined at one place and thus giving us the ability to pin the version if required - without having to look through all the scripts and potentially missing a duplicate. 

Fir libatlas and libjemalloc, feel free to move them to core as they are definitely no api specific dependencies. 

curl, unzip and wget are required by some install scripts and the runtime as well if I'm not mistake. They're also a core dependency.

marcoabreu commented on ci/docker/install/ubuntu_docs.sh:
```diff
@@ -21,8 +21,44 @@
-apt-get install -y doxygen libatlas-base-dev graphviz pandoc
-pip install sphinx==1.3.5 CommonMark==0.5.4 breathe mock recommonmark pypandoc beautifulsoup4
+echo 'Installing Scala...'
+# Setup Scala
```
Scala should already be installed

aaronmarkham replied:
My Scala setup was slightly different... and tested, so I think we should go with mine.

marcoabreu replied:
The problem is that this script is supposed to only install things that are related to the document generation itself. Scala, for example, is installed in https://github.com/apache/incubator-mxnet/blob/master/ci/docker/Dockerfile.build.ubuntu_cpu#L31 and no additional installation should be made at a later point in time. If there's something missing here, it's best to make the change to the install/ubuntu_scala.sh file instead.

At the moment, your commands are overlapping with the scala install script and could thus cause conflicts or unexpected side effects

aaronmarkham replied:
`maven` is already installed, but not `sbt` and `scala`. Do you want me to add those to `ubuntu_scala.sh`?

marcoabreu replied:
Wait, that's actually a good point! Where is the scala runtime actually installed? :O

marcoabreu commented on ci/docker/install/ubuntu_docs.sh:
```diff
@@ -21,8 +21,21 @@
+    recommonmark==0.4.0 \
+    sphinx==1.5.6
+
+echo 'Dependency installation complete.'
```
This looks perfect now, thanks a lot!

marcoabreu commented on docs/build_version_doc/build_all_version.sh:
```diff
@@ -59,16 +61,21 @@ for tag in $tag_list; do
-    git submodule update || exit 1
-    git submodule update --init --recursive
+
+    # uncomment this if you must build in a Python 3 environment
```
Could we remove unused code?

aaronmarkham replied:
Well, that section solves a really annoying problem with building in Python 3 for v0.12.0 or 0.11.0. However, the scala namespace change that just went in breaks that solution. I am figuring that 99% of time most operations will build the latest versions, or it can build the old versions in Python 2, but, if that 1% pops up, you know what to do.

aaronmarkham replied:
I guess I could move that to troubleshooting docs if you want it out of there.

marcoabreu replied:
I think this is related to our previous discussion about using the build scripts of each branch individually instead of making the master script backwards compatible, right?

marcoabreu commented:
Thanks a lot for the quick iteration, Aaron!

Would you mind also changing the main Jenkinsfiles doc generation stage to use your script instead? You can throw out all other branches and make the tag-list only be that PR.

marcoabreu commented:
For the ```ln -s /usr/lib/libopenblas.so /usr/lib/libcblas.so``` error. You can remove that line since cblas is being installed with atlas and thus makes that part obsolete.

aaronmarkham commented:
Currently testing the reorg of the install scripts based on the comments.

1. Removed `libatlas-base-dev` from caffe and moved it to core. Alphabetized the list.
2. Moved a few things to core. Alphabetized the list.
3. Moved `sbt` and `scala` to scala.
3. Cleaned out any dups from docs.

I noticed onnx for the most part duplicates caffe, but made no effort to remedy that. Seems to me like maybe we should at least keep the required deps in the comments, just in case... for example, some removes atlas from core and breaks both docs and caffe because they didn't know better.

marcoabreu commented:
They'll notice pretty quickly if they removed a dependency because our CI won't work in that case and the PR won't get through :)

I think a proper dependency documentation should be at a more prominent place and not in the CI.

aaronmarkham commented:
Are we good on the changes for now?

marcoabreu commented on ci/docker/install/ubuntu_scala.sh:
```diff
@@ -22,10 +22,17 @@
+apt-get update && apt-get install -y \
+    maven \
+    sbt \
+    scala
```
Could we pin the scala version?

aaronmarkham replied:
I'd like to pin everything, not just scala, but let's put that out of scope for now.

marcoabreu commented on docs/Jenkinsfile:
```diff
@@ -0,0 +1,82 @@
+// command to start a docker container
+docker_run = 'tests/ci_build/ci_build.sh'
+// timeout in minutes
+max_time = 1440
```
Set to 60

marcoabreu commented on docs/Jenkinsfile:
```diff
@@ -0,0 +1,82 @@
+// See documents at https://jenkins.io/doc/book/pipeline/jenkinsfile/
+
+// command to start a docker container
+docker_run = 'tests/ci_build/ci_build.sh'
```
Remove this

ThomasDelteil commented:
Awesome job @aaronmarkham ! Does that mean the site will get updated nightly? I realize the current website is a few weeks/months old?

aaronmarkham commented:
Ya, @ThomasDelteil ! 
Nightlies!

ThomasDelteil commented:
@aaronmarkham re the "fixing scala issue" commit, we don't need to install the java sdk to run scala compile docs? Or does it come as a dependency of scala itself?

aaronmarkham commented:
Well, if you run the deps yourself you don't need to install the jdk separately. Just install sbt Scala and maven.

I was messing with that because Oracle was their jdk package down and CI was failing. But removing it didn't work because CI was caching in some way that wasn't picking up my addition of maven which would allow the Scala PKG to be made.

aaronmarkham commented:
BTW, this isn't passing now because of some flaky GPU test.

piiswrong commented on docs/build_version_doc/artifacts/0.12.0.index.html:
```diff
@@ -0,0 +1,1241 @@
+<!DOCTYPE html>
```
what are these?

aaronmarkham replied:
Those patch the install pages to have the pip install version number. The automated route with beautiful soup is broken. The script copies each version into their respective folders. 

We'll get rid of these when we can refactor the install page.

aaronmarkham replied:
Also, note that I added usage instructions in the main comments.

ThomasDelteil commented on ci/docker/install/ubuntu_scala.sh:
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+  #!/bin/bash
```
extra space here shouldn't be here
